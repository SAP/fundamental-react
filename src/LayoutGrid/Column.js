import classnamesBind from 'classnames/bind';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import { mapSize, mapSizeClasses, resolveSpan, validSpan } from './_layoutGridUtils';
import styles from 'fundamental-styles/dist/layout-grid.css';

const classnames = classnamesBind.bind(styles);

const Column = React.forwardRef(({ children, className, cssNamespace, full, offset, offsetPosition, span, ...props }, ref) => {

    const offsets = mapSize((size) => resolveSpan(offset, size, false));

    const offsetPositionModifier = offsetPosition === 'after' ? '-after' : '';

    const responsiveColClasses = mapSizeClasses((size) => {
        if (validSpan(span)) return span;
        if (span && validSpan(span[size])) return span[size];
        return false;
    }, cssNamespace);

    const colClasses = classnames(
        `${cssNamespace}-col`,
        responsiveColClasses,
        {
            [`${cssNamespace}-col--full`]: full,
            [`${cssNamespace}-col--offset${offsetPositionModifier}-${offsets.smallScreen}`]: offsets.smallScreen,
            [`${cssNamespace}-col-md--offset${offsetPositionModifier}-${offsets.mediumScreen}`]: offsets.mediumScreen,
            [`${cssNamespace}-col-lg--offset${offsetPositionModifier}-${offsets.largeScreen}`]: offsets.largeScreen,
            [`${cssNamespace}-col-xl--offset${offsetPositionModifier}-${offsets.xLargeScreen}`]: offsets.xLargeScreen
        },
        className
    );

    return (
        <div
            {...props}
            className={colClasses}
            ref={ref}>
            {children}
        </div>
    );
});

Column.displayName = 'Column';

Column.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to `true` to have the column automatically occupy the remaining space in the row */
    full: PropTypes.bool,
    /** How many cells out of 12 should the column be offset by on each screen size. Defaults to none.
     *
     * Example usage:
     *
     * * `offset={4}` -for all screen sizes
     * * `offset={{ smallScreen: 8}}` -for small screen sizes only, none for other sizes
     * * `offset={{ smallScreen: 6, xLargeScreen: 1}}` -so on
     * * `offset={{ smallScreen: 6, mediumScreen: 5, largeScreen: 3, xLargeScreen: 1}}` -so forth
    */
    offset: CustomPropTypes.validColumnProp,
    /** Are the offsets to be applied `before` or `after` the column? */
    offsetPosition: PropTypes.oneOf(['after', 'before']),
    /** How many cells out of 12 should the column occupy on each screen size. Defaults to 12.
     *
     * Example usage:
     *
     * * `span={4}` -for all screen sizes
     * * `span={{ smallScreen: 8}}` -for small screen sizes only, none for other sizes
     * * `span={{ smallScreen: 6, xLargeScreen: 1}}` -so on
     * * `span={{ smallScreen: 6, mediumScreen: 5, largeScreen: 3, xLargeScreen: 1}}` -so forth
    */
    span: CustomPropTypes.validColumnProp
};

Column.defaultProps = {
    offsetPosition: 'before'
};


export default withStyles(Column);
