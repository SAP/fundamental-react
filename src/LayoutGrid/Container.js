import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/layout-grid.css';

const classnames = classnamesBind.bind(styles);

/** **Container** can be used to arrange wrap the **Row** and **Column** components to create a responsive grid layout with appropriate padding.*/
const Container = React.forwardRef(({ noGap, children, className, cssNamespace, ...props }, ref) => {

    const containerClasses = classnames(
        `${cssNamespace}-container`,
        {
            [`${cssNamespace}-container--no-gap`]: noGap
        },
        className
    );

    return (
        <div
            {...props}
            className={containerClasses}
            ref={ref}>
            {children}
        </div>
    );
});

Container.displayName = 'Container';

Container.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to remove the margins between the panels */
    noGap: PropTypes.bool
};

export default withStyles(Container);
