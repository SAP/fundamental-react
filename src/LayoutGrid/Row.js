/* eslint-disable no-console */
import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import { calculateSpan, getNodeSpan, hasSpan, mapSize } from './_layoutGridUtils';
import styles from 'fundamental-styles/dist/layout-grid.css';

const classnames = classnamesBind.bind(styles);

const Row = React.forwardRef(({ children, className, cssNamespace, ...props }, ref) => {

    const rowClasses = classnames(
        `${cssNamespace}-row`,
        className
    );
    let occupiedSpans = mapSize(() => 0);
    let unspannedChildren = mapSize(() => 0);


    React.Children.forEach(children, (child) => {
        occupiedSpans = mapSize((size) => occupiedSpans[size] + getNodeSpan(child, size));
        unspannedChildren = mapSize((size) => unspannedChildren[size] + (hasSpan(child, size) ? 0 : 1));
    });

    let availableSpans = mapSize((size) => 12 - (( -1 < occupiedSpans[size] && occupiedSpans[size] < 13 ) ? occupiedSpans[size] : 0));

    let spannedChildren = React.Children.map(children, (child) => {
        const childSpan = mapSize((size) => calculateSpan(child, availableSpans, unspannedChildren, size));
        return React.cloneElement(child, {
            span: childSpan
        });
    });


    return (
        <div
            {...props}
            className={rowClasses}
            ref={ref}>
            {spannedChildren}
        </div>
    );
});

Row.displayName = 'Row';

Row.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string
};

export default withStyles(Row);
