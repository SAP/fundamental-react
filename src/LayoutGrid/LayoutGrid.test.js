import Column from './Column';
import Container from './Container';
import { mount } from 'enzyme';
import React from 'react';
import Row from './Row';
import {
    toHaveClass
} from '@testing-library/jest-dom/matchers';

expect.extend({ toHaveClass });

describe('<Row />', () => {
    describe('Auto spanning', () => {
        test('Row should automatically assign spans to un-spanned Columns', () => {
            let wrapper = mount(
                <Container>
                    <Row>
                        <Column
                            id='firstColumn'
                            span={{
                                smallScreen: 12,
                                mediumScreen: 8,
                                largeScreen: 7,
                                xLargeScreen: 6
                            }}>
                            We have defined col spans for this Column
                        </Column>
                        <Column
                            id='secondColumn' >
                            This Column will share available cells
                        </Column>
                        <Column
                            id='thirdColumn' >
                            This Column will share available cells
                        </Column>
                    </Row>
                </Container>
            );
            expect(wrapper.find('div#firstColumn').getDOMNode()).toHaveClass('fd-col fd-col--12 fd-col-md--8 fd-col-lg--7 fd-col-xl--6');
            expect(wrapper.find('div#secondColumn').getDOMNode()).toHaveClass('fd-col fd-col--12 fd-col-md--2 fd-col-lg--3 fd-col-xl--3');
            expect(wrapper.find('div#secondColumn').getDOMNode()).toHaveClass('fd-col fd-col--12 fd-col-md--2 fd-col-lg--3 fd-col-xl--3');
            wrapper.unmount();
        });
    });

    test('Alls refs are forwarded the ref', () => {
        let containerRef;
        let rowRef;
        let columnRef;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                containerRef = React.createRef();
                rowRef = React.createRef();
                columnRef = React.createRef();
            }
            render = () => (
                <Container ref={containerRef}>
                    <Row ref={rowRef}>
                        <Column ref={columnRef}>
                            Content
                        </Column>
                    </Row>
                </Container>
            );
        }
        mount(<Test />);
        expect(containerRef.current.tagName).toEqual('DIV');
        expect(containerRef.current).toHaveClass('fd-container');

        expect(rowRef.current.tagName).toEqual('DIV');
        expect(rowRef.current).toHaveClass('fd-row');

        expect(columnRef.current.tagName).toEqual('DIV');
        expect(columnRef.current).toHaveClass('fd-col');
    });
});
