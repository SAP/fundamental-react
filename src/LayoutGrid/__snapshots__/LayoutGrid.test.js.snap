// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayoutGrid /> layoutGrid snapshots 1`] = `
<div
  className="fd-layout-grid blue"
>
  <div
    className="fd-panel fd-layout-grid__span-column-2"
  >
    <div
      className="fd-panel__head"
    >
      <h3
        className="fd-panel__title"
      >
        Panel Header with Actions
      </h3>
    </div>
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__head"
    >
      <p
        className="fd-panel__description"
      >
        Panel Description
      </p>
    </div>
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__head"
    />
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
</div>
`;

exports[`<LayoutGrid /> layoutGrid snapshots 2`] = `
<div
  className="fd-layout-grid fd-layout-grid--col-2"
>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
</div>
`;

exports[`<LayoutGrid /> layoutGrid snapshots 3`] = `
<div
  className="fd-layout-grid fd-layout-grid--nogap"
>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
  <div
    className="fd-panel"
  >
    <div
      className="fd-panel__body"
    >
      Panel
    </div>
  </div>
</div>
`;
